kotlinx-coroutines-core/common/src/internal/DispatchedTask.kt

```diff
@@ -8,12 +8,17 @@ import kotlinx.coroutines.internal.*
 import kotlin.coroutines.*
 import kotlin.jvm.*
 
-@PublishedApi internal const val MODE_ATOMIC_DEFAULT = 0 // schedule non-cancellable dispatch for suspendCoroutine
-@PublishedApi internal const val MODE_CANCELLABLE = 1    // schedule cancellable dispatch for suspendCancellableCoroutine
-@PublishedApi internal const val MODE_UNDISPATCHED = 2   // when the thread is right, but need to mark it with current coroutine
+@PublishedApi
+internal const val MODE_ATOMIC_DEFAULT = 0 // schedule non-cancellable dispatch for suspendCoroutine
+@PublishedApi
+internal const val MODE_CANCELLABLE = 1    // schedule cancellable dispatch for suspendCancellableCoroutine
 
-internal val Int.isCancellableMode get() = this == MODE_CANCELLABLE
-internal val Int.isDispatchedMode get() = this == MODE_ATOMIC_DEFAULT || this == MODE_CANCELLABLE
+internal const val MODE_CANCELLABLE_REUSABLE = 2 // same as MODE_CANCELLABLE but supports reused
```

Comment:
I am a bit concerned about the maintainability of this machinery.
E.g. it is easy to use `if (mode == MODE_CANCELLABLE) doSomeLogic()` instead of `mode.isCancellableMode`.
Let's at least leave a precautious comment with a suggestion to avoid using a comparison with constants directly.

Reply:
>E.g. it is easy to use if (mode == MODE_CANCELLABLE) doSomeLogic() instead of mode.isCancellableMode.

Speaking of which, please take a look at `CancellableContinuationImpl#getResult:237`

Reply:
Good catch. Added test for that particular omission.